crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+        }
+        Ok(())
+    }
+
+    /// Collect bundles within the new blocks of the best chain, blocks are handled from
+    /// older blcok to newer blcok, an `Err` may return in the middle and left some blocks
+    /// unhandled, these blocks will be handled when processing the next new best block.
+    fn on_import_new_best_block(
+        &mut self,
+        new_best_hash: Block::Hash,
+    ) -> Result<(), sp_blockchain::Error> {
+        let current_best_hash = match self.best_hash() {
+            Some(h) => h,
+            // Collect bundle from the last K blocks if it have not process any block
+            None => {
+                self.collect_bundle_from_last_k(new_best_hash)?;
+                return Ok(());
+            }
+        };
+
+        let route = sp_blockchain::tree_route(&*self.client, current_best_hash, new_best_hash)?;
+        let (retracted, enacted) = (route.retracted(), route.enacted());
+
+        // Remove bundles from the stale fork
+        for retracted_block in retracted {
+            match self.bundle_stored_in_last_k.front() {
+                Some((block_hash, _)) if *block_hash == retracted_block.hash => {
+                    self.bundle_stored_in_last_k.pop_front();
+                }
+                bb => {
+                    return Err(sp_blockchain::Error::Application(Box::from(
+                        format!(
+                            "Got wrong block from the bundle-collector, expect {:?}, got {:?}, this should not happen",
+                            retracted_block,
+                            bb.map(|(block_hash, _)| block_hash),
+                        ),
+                    )));
+                }
+            }
+        }
+
+        // Add bundles from the new block of the best fork
+        for enacted_block in enacted {
+            let block_bundles = self.collect_bundle_from_block(enacted_block.hash)?;
+            self.bundle_stored_in_last_k.push_front(block_bundles);
+        }
+
+        // Remove blocks from the back end to keep at most the bundle of the last K blocks
+        self.bundle_stored_in_last_k.truncate(self.confirm_depth_k);
+
+        Ok(())
+    }
+
+    pub async fn update_recent_stored_bundles(&mut self, new_best_hash: Block::Hash) {
+        let pre_best_hash = self.best_hash();
+        if let Err(err) = self.on_import_new_best_block(new_best_hash) {
+            tracing::error!(
+                %err,
+                "Failed to handle new best block for bundle-collector"
+            );
+        }
+        if pre_best_hash != self.best_hash() {
+            self.bundle_syncer
+                .update(self.bundle_stored_in_last_k.clone());
+        }
+    }
+}
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
+    #[allow(clippy::type_complexity)]
+    bundle_stored_in_last_k: Arc<Mutex<VecDeque<(Block::Hash, HashSet<Hash>)>>>,
```

Comment:
If multiple concurrent readers(validators) are possible, RwLock would be a better fit here